File: orgo_mcp/tools/streaming.py
"""RTMP streaming tools — stream computer display to Twitch, YouTube, etc.

Requires pre-configured RTMP connections in your Orgo account settings.
"""

import json

from orgo_mcp.server import mcp
from orgo_mcp.auth import get_current_api_key
from orgo_mcp.client import computer_action
from orgo_mcp.errors import handle_orgo_error
from orgo_mcp.models import StartStreamInput, ComputerIdInput


@mcp.tool(
    name="orgo_start_stream",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": False, "openWorldHint": True},
)
async def orgo_start_stream(params: StartStreamInput) -> str:
    """Start RTMP streaming from a computer to a pre-configured connection. One stream per computer."""
    try:
        api_key = get_current_api_key(mcp)
        data = await computer_action(
            "POST", params.computer_id, "stream/start", api_key,
            json={"connection_name": params.connection_name},
        )
        return json.dumps(data, indent=2)
    except Exception as e:
        return handle_orgo_error(e)


@mcp.tool(
    name="orgo_stream_status",
    annotations={"readOnlyHint": True, "destructiveHint": False, "idempotentHint": True, "openWorldHint": True},
)
async def orgo_stream_status(params: ComputerIdInput) -> str:
    """Get current streaming status: idle, streaming, or terminated."""
    try:
        api_key = get_current_api_key(mcp)
        data = await computer_action("GET", params.computer_id, "stream/status", api_key)
        return json.dumps(data, indent=2)
    except Exception as e:
        return handle_orgo_error(e)


@mcp.tool(
    name="orgo_stop_stream",
    annotations={"readOnlyHint": False, "destructiveHint": False, "idempotentHint": True, "openWorldHint": True},
)
async def orgo_stop_stream(params: ComputerIdInput) -> str:
    """Stop an active RTMP stream."""
    try:
        api_key = get_current_api_key(mcp)
        data = await computer_action("POST", params.computer_id, "stream/stop", api_key)
        return json.dumps(data, indent=2)
    except Exception as e:
        return handle_orgo_error(e)

File: src/tools/streaming.ts
/**
 * RTMP streaming tools -- stream computer display to Twitch, YouTube, etc.
 *
 * Requires pre-configured RTMP connections in your Orgo account settings.
 */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { getApiKey, resolveComputerId } from "../auth.js";
import { computerAction } from "../client.js";
import { handleError } from "../errors.js";

export function registerStreamingTools(server: McpServer): void {
  server.tool(
    "orgo_start_stream",
    "Start RTMP streaming from a computer to a pre-configured connection. One stream per computer.",
    {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
      connection_name: z.string().min(1).describe("Name of pre-configured RTMP connection (set up at orgo.ai)"),
    },
    async ({ computer_id, connection_name }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        const data = await computerAction("POST", id, "stream/start", apiKey, {
          json: { connection_name },
        });
        return { content: [{ type: "text" as const, text: JSON.stringify(data, null, 2) }] };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    }
  );

  server.tool(
    "orgo_stream_status",
    "Get current streaming status: idle, streaming, or terminated.",
    {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
    },
    async ({ computer_id }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        const data = await computerAction("GET", id, "stream/status", apiKey);
        return { content: [{ type: "text" as const, text: JSON.stringify(data, null, 2) }] };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    }
  );

  server.tool(
    "orgo_stop_stream",
    "Stop an active RTMP stream.",
    {
      computer_id: z.string().optional().describe("Computer ID (uses ORGO_DEFAULT_COMPUTER_ID if omitted)"),
    },
    async ({ computer_id }) => {
      try {
        const apiKey = getApiKey();
        const id = resolveComputerId(computer_id);
        const data = await computerAction("POST", id, "stream/stop", apiKey);
        return { content: [{ type: "text" as const, text: JSON.stringify(data, null, 2) }] };
      } catch (e) {
        return { content: [{ type: "text" as const, text: handleError(e) }], isError: true };
      }
    }
  );
}
